source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,197 @@
+#include "common/router/filter_config_discovery_impl.h"
+
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Router {
+
+DynamicFilterConfigProviderImpl::DynamicFilterConfigProviderImpl(
+    FilterConfigSubscriptionSharedPtr&& subscription, bool require_terminal,
+    Server::Configuration::FactoryContext& factory_context)
+    : subscription_(std::move(subscription)), require_terminal_(require_terminal),
+      tls_(factory_context.threadLocal().allocateSlot()),
+      init_target_("DynamicFilterConfigProviderImpl", [this]() {
+        subscription_->start();
+        init_target_.ready();
+      }) {
+  subscription_->filter_config_providers_.insert(this);
+  tls_->set([](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    return std::make_shared<ThreadLocalConfig>();
+  });
+}
+
+DynamicFilterConfigProviderImpl::~DynamicFilterConfigProviderImpl() {
+  subscription_->filter_config_providers_.erase(this);
+}
+
+const std::string& DynamicFilterConfigProviderImpl::name() { return subscription_->name(); }
+
+absl::optional<Http::FilterFactoryCb> DynamicFilterConfigProviderImpl::config() {
+  return tls_->getTyped<ThreadLocalConfig>().config_;
+}
+
+void DynamicFilterConfigProviderImpl::validateConfig(
+    Server::Configuration::NamedHttpFilterConfigFactory& factory) {
+  if (factory.isTerminalFilter() && !require_terminal_) {
```

Comment:
Why are we validating this down here? I'd have thought it would be sufficient to validate the terminal filter condition in HCM..

Reply:
Config is rejected if it cannot be applied. Right now it's just terminal checks, but I imagine there'll be more. We can restructure the code to use callbacks if it's more clear that way.

Reply:
I'm still not sure where this additional config validation requirement comes from, in particular as a first-class API feature?